arch/arm/core/aarch32/cortex_m/thread_abort.c

```diff
@@ -24,29 +24,29 @@
 #include <wait_q.h>
 #include <sys/__assert.h>
 
-extern void z_thread_single_abort(struct k_thread *thread);
-
 void z_impl_k_thread_abort(k_tid_t thread)
 {
-	unsigned int key;
-
-	key = irq_lock();
-
-	__ASSERT(!(thread->base.user_options & K_ESSENTIAL),
-		 "essential thread aborted");
-
 	z_thread_single_abort(thread);
-	z_thread_monitor_exit(thread);
 
 	if (_current == thread) {
-		if ((SCB->ICSR & SCB_ICSR_VECTACTIVE_Msk) == 0) {
-			(void)z_swap_irqlock(key);
-			CODE_UNREACHABLE;
-		} else {
+		if (arch_is_in_isr()) {
+			/* ARM is unlike most arches in that this is true
+			 * even for non-peripheral interrupts, even though
+			 * for these types of faults there is not an implicit
+			 * reschedule on the way out. See #21923.
+			 *
+			 * We have to reschedule since the current thread
+			 * should no longer run after we return, so
+			 * Trigger PendSV, in case we are in one of the
+			 * situations where the isr check is true but there
+			 * is not an implicit scheduler invocation.
```

Comment:
@andrewboie i guess this could be solved by having 2 APIs:
- arch_is_in_hw_isr()
- arch_is_in_sys_exception()

Reply:
yeah thinking along these lines. I am going to try to carve out some time for a proposal on that ticket sometime in 2.5